doc/man/scons.1
.\" __COPYRIGHT__
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining
.\" a copy of this software and associated documentation files (the
.\" "Software"), to deal in the Software without restriction, including
.\" without limitation the rights to use, copy, modify, merge, publish,
.\" distribute, sublicense, and/or sell copies of the Software, and to
.\" permit persons to whom the Software is furnished to do so, subject to
.\" the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included
.\" in all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
.\" KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
.\" WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
.\" NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
.\" LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
.\" OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
.\" WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
.\"
.\" __FILE__ __REVISION__ __DATE__ __DEVELOPER__
.\"
.TH SCONS 1 "__MONTH_YEAR__"
.\" ES - Example Start - indents and turns off line fill
.rm ES
.de ES
.RS
.nf
..
.\" EE - Example End - ends indent and turns line fill back on
.rm EE
.de EE
.fi
.RE
..
.SH NAME
scons \- a software construction tool
.SH SYNOPSIS
.B scons
[
.IR options ...
]
[
.IR name = val ...
]
[
.IR targets ...
]
.SH DESCRIPTION

The
.B scons
utility builds software (or other files) by determining which
component pieces must be rebuilt and executing the necessary commands to
rebuild them.

By default,
.B scons
searches for a file named
.IR SConstruct ,
.IR Sconstruct ,
or
.I sconstruct
(in that order) in the current directory and reads its
configuration from the first file found.
An alternate file name may be
specified via the
.B -f
option.

The
.I SConstruct
file can specify subsidiary
configuration files using the
.BR SConscript ()
function.
By convention,
these subsidiary files are named
.IR SConscript ,
although any name may be used.
(Because of this naming convention,
the term "SConscript files"
is sometimes used to refer
generically to all
.B scons
configuration files,
regardless of actual file name.)

The configuration files
specify the target files to be built, and
(optionally) the rules to build those targets.  Reasonable default
rules exist for building common software components (executable
programs, object files, libraries), so that for most software
projects, only the target and input files need be specified.

Before reading the
.I SConstruct
file,
.B scons
looks for a directory named
.I site_scons
in various system directories (see below) and the directory containing the
.I SConstruct
file; for each of those dirs which exists,
.I site_scons
is prepended to sys.path,
the file
.IR site_scons/site_init.py ,
is evaluated if it exists,
and the directory
.I site_scons/site_tools
is prepended to the default toolpath if it exists.
See the
.I --no-site-dir
and
.I --site-dir
options for more details.

.B scons
reads and executes the SConscript files as Python scripts,
so you may use normal Python scripting capabilities
(such as flow control, data manipulation, and imported Python libraries)
to handle complicated build situations.
.BR scons ,
however, reads and executes all of the SConscript files
.I before
it begins building any targets.
To make this obvious,
.B scons
prints the following messages about what it is doing:

.ES
$ scons foo.out
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets  ...
cp foo.in foo.out
scons: done building targets.
$
.EE

The status messages
(everything except the line that reads "cp foo.in foo.out")
may be suppressed using the
.B -Q
option.

.B scons
does not automatically propagate
the external environment used to execute
.B scons
to the commands used to build target files.
This is so that builds will be guaranteed
repeatable regardless of the environment
variables set at the time
.B scons
is invoked.
This also means that if the compiler or other commands
that you want to use to build your target files
are not in standard system locations,
.B scons
will not find them unless
you explicitly set the PATH
to include those locations.
Whenever you create an
.B scons
construction environment,
you can propagate the value of PATH
from your external environment as follows:

.ES
import os
env = Environment(ENV = {'PATH' : os.environ['PATH']})
.EE

Similarly, if the commands use external environment variables
like $PATH, $HOME, $JAVA_HOME, $LANG, $SHELL, $TERM, etc.,
these variables can also be explicitly propagated:

.ES
import os
env = Environment(ENV = {'PATH' : os.environ['PATH'],
                         'HOME' : os.environ['HOME']})
.EE

Or you may explicitly propagate the invoking user's
complete external environment:

.ES
import os
env = Environment(ENV = os.environ)
.EE

This comes at the expense of making your build
dependent on the user's environment being set correctly,
but it may be more convenient for many configurations.

.B scons
can scan known input files automatically for dependency
information (for example, #include statements
in C or C++ files) and will rebuild dependent files appropriately
whenever any "included" input file changes.
.B scons
supports the
ability to define new scanners for unknown input file types.

.B scons
knows how to fetch files automatically from
SCCS or RCS subdirectories
using SCCS, RCS or BitKeeper.

.B scons
is normally executed in a top-level directory containing a
.I SConstruct
file, optionally specifying
as command-line arguments
the target file or files to be built.

By default, the command

.ES
scons
.EE

will build all target files in or below the current directory.
Explicit default targets
(to be built when no targets are specified on the command line)
may be defined the SConscript file(s)
using the
.B Default()
function, described below.

Even when
.B Default()
targets are specified in the SConscript file(s),
all target files in or below the current directory
may be built by explicitly specifying
the current directory (.)
as a command-line target:

.ES
scons .
.EE

Building all target files,
including any files outside of the current directory,
may be specified by supplying a command-line target
of the root directory (on POSIX systems):

.ES
scons /
.EE

or the path name(s) of the volume(s) in which all the targets
should be built (on Windows systems):

.ES
scons C:\\ D:\\
.EE

To build only specific targets,
supply them as command-line arguments:

.ES
scons foo bar
.EE

in which case only the specified targets will be built
(along with any derived files on which they depend).

Specifying "cleanup" targets in SConscript files is not usually necessary.
The
.B -c
flag removes all files
necessary to build the specified target:

.ES
scons -c .
.EE

to remove all target files, or:

.ES
scons -c build export
.EE

to remove target files under build and export.
Additional files or directories to remove can be specified using the
.BR Clean()
function.
Conversely, targets that would normally be removed by the
.B -c
invocation
can be prevented from being removed by using the
.BR NoClean ()
function.

A subset of a hierarchical tree may be built by
remaining at the top-level directory (where the
.I SConstruct
file lives) and specifying the subdirectory as the target to be
built:

.ES
scons src/subdir
.EE

or by changing directory and invoking scons with the
.B -u
option, which traverses up the directory
hierarchy until it finds the
.I SConstruct
file, and then builds
targets relatively to the current subdirectory:

.ES
cd src/subdir
scons -u .
.EE

.B scons
supports building multiple targets in parallel via a
.B -j
option that takes, as its argument, the number
of simultaneous tasks that may be spawned:

.ES
scons -j 4
.EE

builds four targets in parallel, for example.

.B scons
can maintain a cache of target (derived) files that can
be shared between multiple builds.  When caching is enabled in a
SConscript file, any target files built by
.B scons
will be copied
to the cache.  If an up-to-date target file is found in the cache, it
will be retrieved from the cache instead of being rebuilt locally.
Caching behavior may be disabled and controlled in other ways by the
.BR --cache-force ,
.BR --cache-disable ,
and
.B --cache-show
command-line options.  The
.B --random
option is useful to prevent multiple builds
from trying to update the cache simultaneously.

Values of variables to be passed to the SConscript file(s)
may be specified on the command line:

.ES
scons debug=1 .
.EE

These variables are available in SConscript files
through the ARGUMENTS dictionary,
and can be used in the SConscript file(s) to modify
the build in any way:

.ES
if ARGUMENTS.get('debug', 0):
    env = Environment(CCFLAGS = '-g')
else:
    env = Environment()
.EE

The command-line variable arguments are also available
in the ARGLIST list,
indexed by their order on the command line.
This allows you to process them in order rather than by name,
if necessary.
ARGLIST[0] returns a tuple
containing (argname, argvalue).
A Python exception is thrown if you
try to access a list member that
does not exist.

.B scons
requires Python version 2.4 or later.
There should be no other dependencies or requirements to run
.B scons.

.\" The following paragraph reflects the default tool search orders
.\" currently in SCons/Tool/__init__.py.  If any of those search orders
.\" change, this documentation should change, too.
By default,
.B scons
knows how to search for available programming tools
on various systems.
On Windows systems,
.B scons
searches in order for the
Microsoft Visual C++ tools,
the MinGW tool chain,
the Intel compiler tools,
and the PharLap ETS compiler.
On OS/2 systems,
.B scons
searches in order for the
OS/2 compiler,
the GCC tool chain,
and the Microsoft Visual C++ tools,
On SGI IRIX, IBM AIX, Hewlett Packard HP-UX, and Sun Solaris systems,
.B scons
searches for the native compiler tools
(MIPSpro, Visual Age, aCC, and Forte tools respectively)
and the GCC tool chain.
On all other platforms,
including POSIX (Linux and UNIX) platforms,
.B scons
searches in order
for the GCC tool chain,
the Microsoft Visual C++ tools,
and the Intel compiler tools.
You may, of course, override these default values
by appropriate configuration of
Environment construction variables.

.SH OPTIONS
In general,
.B scons
supports the same command-line options as GNU
.BR make ,
and many of those supported by
.BR cons .

.TP
-b
Ignored for compatibility with non-GNU versions of
.BR make.

.TP
-c, --clean, --remove
Clean up by removing all target files for which a construction
command is specified.
Also remove any files or directories associated to the construction command
using the
.BR Clean ()
function.
Will not remove any targets specified by the
.BR NoClean ()
function.

.TP
.RI --cache-debug= file
Print debug information about the
.BR CacheDir ()
derived-file caching
to the specified
.IR file .
If
.I file
is
.B \-
(a hyphen),
the debug information are printed to the standard output.
The printed messages describe what signature file names are
being looked for in, retrieved from, or written to the
.BR CacheDir ()
directory tree.

.TP
--cache-disable, --no-cache
Disable the derived-file caching specified by
.BR CacheDir ().
.B scons
will neither retrieve files from the cache
nor copy files to the cache.

.TP
--cache-force, --cache-populate
When using
.BR CacheDir (),
populate a cache by copying any already-existing, up-to-date
derived files to the cache,
in addition to files built by this invocation.
This is useful to populate a new cache with
all the current derived files,
or to add to the cache any derived files
recently built with caching disabled via the
.B --cache-disable
option.

.TP
--cache-show
When using
.BR CacheDir ()
and retrieving a derived file from the cache,
show the command
that would have been executed to build the file,
instead of the usual report,
"Retrieved `file' from cache."
This will produce consistent output for build logs,
regardless of whether a target
file was rebuilt or retrieved from the cache.

.TP
.RI --config= mode
This specifies how the
.B Configure
call should use or generate the
results of configuration tests.
The option should be specified from
among the following choices:

.TP
--config=auto
scons will use its normal dependency mechanisms
to decide if a test must be rebuilt or not.
This saves time by not running the same configuration tests
every time you invoke scons,
but will overlook changes in system header files
or external commands (such as compilers)
if you don't specify those dependecies explicitly.
This is the default behavior.

.TP
--config=force
If this option is specified,
all configuration tests will be re-run
regardless of whether the
cached results are out of date.
This can be used to explicitly
force the configuration tests to be updated
in response to an otherwise unconfigured change
in a system header file or compiler.

.TP
--config=cache
If this option is specified,
no configuration tests will be rerun
and all results will be taken from cache.
Note that scons will still consider it an error
if --config=cache is specified
and a necessary test does not
yet have any results in the cache.

.TP
.RI "-C" " directory" ",  --directory=" directory
Change to the specified
.I directory
before searching for the
.IR SConstruct ,
.IR Sconstruct ,
or
.I sconstruct
file, or doing anything
else.  Multiple
.B -C
options are interpreted
relative to the previous one, and the right-most
.B -C
option wins. (This option is nearly
equivalent to
.BR "-f directory/SConstruct" ,
except that it will search for
.IR SConstruct ,
.IR Sconstruct ,
or
.I sconstruct
in the specified directory.)

.\" .TP
.\" -d
.\" Display dependencies while building target files.  Useful for
.\" figuring out why a specific file is being rebuilt, as well as
.\" general debugging of the build process.

.TP
-D
Works exactly the same way as the
.B -u
option except for the way default targets are handled.
When this option is used and no targets are specified on the command line,
all default targets are built, whether or not they are below the current
directory.

.TP
.RI --debug= type
Debug the build process.
.I type
specifies what type of debugging:

.TP
--debug=count
Print how many objects are created
of the various classes used internally by SCons
before and after reading the SConscript files
and before and after building targets.
This is not supported when SCons is executed with the Python
.B -O
(optimized) option
or when the SCons modules
have been compiled with optimization
(that is, when executing from
.B *.pyo
files).

.TP
--debug=duplicate
Print a line for each unlink/relink (or copy) of a variant file from
its source file.  Includes debugging info for unlinking stale variant
files, as well as unlinking old targets before building them.

.TP
--debug=dtree
A synonym for the newer
.B --tree=derived
option.
This will be deprecated in some future release
and ultimately removed.

.TP
--debug=explain
Print an explanation of precisely why
.B scons
is deciding to (re-)build any targets.
(Note:  this does not print anything
for targets that are
.I not
rebuilt.)

.TP
--debug=findlibs
Instruct the scanner that searches for libraries
to print a message about each potential library
name it is searching for,
and about the actual libraries it finds.

.TP
--debug=includes
Print the include tree after each top-level target is built.
This is generally used to find out what files are included by the sources
of a given derived file:

.ES
$ scons --debug=includes foo.o
.EE

.TP
--debug=memoizer
Prints a summary of hits and misses using the Memoizer,
an internal subsystem that counts
how often SCons uses cached values in memory
instead of recomputing them each time they're needed.

.TP
--debug=memory
Prints how much memory SCons uses
before and after reading the SConscript files
and before and after building targets.

.TP
--debug=nomemoizer
A deprecated option preserved for backwards compatibility.

.TP
--debug=objects
Prints a list of the various objects
of the various classes used internally by SCons.

.TP
--debug=pdb
Re-run SCons under the control of the
.RI pdb
Python debugger.

.TP
--debug=prepare
Print a line each time any target (internal or external)
is prepared for building.
.B scons
prints this for each target it considers, even if that
target is up to date (see also --debug=explain).
This can help debug problems with targets that aren't being
built; it shows whether
.B scons
is at least considering them or not.

.TP
--debug=presub
Print the raw command line used to build each target
before the construction environment variables are substituted.
Also shows which targets are being built by this command.
Output looks something like this:
.ES
$ scons --debug=presub
Building myprog.o with action(s):
  $SHCC $SHCFLAGS $SHCCFLAGS $CPPFLAGS $_CPPINCFLAGS -c -o $TARGET $SOURCES
\&...
.EE

.TP
--debug=stacktrace
Prints an internal Python stack trace
when encountering an otherwise unexplained error.

.TP
--debug=stree
A synonym for the newer
.B --tree=all,status
option.
This will be deprecated in some future release
and ultimately removed.

.TP
--debug=time
Prints various time profiling information:
the time spent executing each individual build command;
the total build time (time SCons ran from beginning to end);
the total time spent reading and executing SConscript files;
the total time spent SCons itself spend running
(that is, not counting reading and executing SConscript files);
and both the total time spent executing all build commands
and the elapsed wall-clock time spent executing those build commands.
(When
.B scons
is executed without the
.B -j
option,
the elapsed wall-clock time will typically
be slightly longer than the total time spent
executing all the build commands,
due to the SCons processing that takes place
in between executing each command.
When
.B scons
is executed
.I with
the
.B -j
option,
and your build configuration allows good parallelization,
the elapsed wall-clock time should
be significantly smaller than the
total time spent executing all the build commands,
since multiple build commands and
intervening SCons processing
should take place in parallel.)

.TP
--debug=tree
A synonym for the newer
.B --tree=all
option.
This will be deprecated in some future release
and ultimately removed.

.TP
.RI --diskcheck= types
Enable specific checks for
whether or not there is a file on disk
where the SCons configuration expects a directory
(or vice versa),
and whether or not RCS or SCCS sources exist
when searching for source and include files.
The
.I types
argument can be set to:
.BR all ,
to enable all checks explicitly
(the default behavior);
.BR none ,
to disable all such checks;
.BR match ,
to check that files and directories on disk
match SCons' expected configuration;
.BR rcs ,
to check for the existence of an RCS source
for any missing source or include files;
.BR sccs ,
to check for the existence of an SCCS source
for any missing source or include files.
Multiple checks can be specified separated by commas;
for example,
.B --diskcheck=sccs,rcs
would still check for SCCS and RCS sources,
but disable the check for on-disk matches of files and directories.
Disabling some or all of these checks
can provide a performance boost for large configurations,
or when the configuration will check for files and/or directories
across networked or shared file systems,
at the slight increased risk of an incorrect build
or of not handling errors gracefully
(if include files really should be
found in SCCS or RCS, for example,
or if a file really does exist
where the SCons configuration expects a directory).

.TP
.RI --duplicate= ORDER
There are three ways to duplicate files in a build tree: hard links,
soft (symbolic) links and copies. The default behaviour of SCons is to
prefer hard links to soft links to copies. You can specify different
behaviours with this option.
.IR ORDER
must be one of
.IR hard-soft-copy
(the default),
.IR soft-hard-copy ,
.IR hard-copy ,
.IR soft-copy
or
.IR copy .
SCons will attempt to duplicate files using
the mechanisms in the specified order.

.\" .TP
.\" -e, --environment-overrides
.\" Variables from the execution environment override construction
.\" variables from the SConscript files.

.TP
.RI -f " file" ", --file=" file ", --makefile=" file ", --sconstruct=" file
Use
.I file
as the initial SConscript file.
Multiple
.B -f
options may be specified,
in which case
.B scons
will read all of the specified files.

.TP
-h, --help
Print a local help message for this build, if one is defined in
the SConscript file(s), plus a line that describes the
.B -H
option for command-line option help.  If no local help message
is defined, prints the standard help message about command-line
options.  Exits after displaying the appropriate message.

.TP
-H, --help-options
Print the standard help message about command-line options and
exit.

.TP
-i, --ignore-errors
Ignore all errors from commands executed to rebuild files.

.TP
.RI -I " directory" ", --include-dir=" directory
Specifies a
.I directory
to search for
imported Python modules.  If several
.B -I
options
are used, the directories are searched in the order specified.

.TP
--implicit-cache
Cache implicit dependencies.
This causes
.B scons
to use the implicit (scanned) dependencies
from the last time it was run
instead of scanning the files for implicit dependencies.
This can significantly speed up SCons,
but with the following limitations:
.IP
.B scons
will not detect changes to implicit dependency search paths
(e.g.
.BR CPPPATH ", " LIBPATH )
that would ordinarily
cause different versions of same-named files to be used.
.IP
.B scons
will miss changes in the implicit dependencies
in cases where a new implicit
dependency is added earlier in the implicit dependency search path
(e.g.
.BR CPPPATH ", " LIBPATH )
than a current implicit dependency with the same name.

.TP
--implicit-deps-changed
Forces SCons to ignore the cached implicit dependencies. This causes the
implicit dependencies to be rescanned and recached. This implies
.BR --implicit-cache .

.TP
--implicit-deps-unchanged
Force SCons to ignore changes in the implicit dependencies.
This causes cached implicit dependencies to always be used.
This implies
.BR --implicit-cache .

.TP
--interactive
Starts SCons in interactive mode.
The SConscript files are read once and a
.B "scons>>>"
prompt is printed.
Targets may now be rebuilt by typing commands at interactive prompt
without having to re-read the SConscript files
and re-initialize the dependency graph from scratch.

SCons interactive mode supports the following commands:

.RS 10
.TP 6
.BI build "[OPTIONS] [TARGETS] ..."
Builds the specified
.I TARGETS
(and their dependencies)
with the specified
SCons command-line
.IR OPTIONS .
.B b
and
.B scons
are synonyms.

The following SCons command-line options affect the
.B build
command:

.ES
--cache-debug=FILE
--cache-disable, --no-cache
--cache-force, --cache-populate
--cache-show
--debug=TYPE
-i, --ignore-errors
-j N, --jobs=N
-k, --keep-going
-n, --no-exec, --just-print, --dry-run, --recon
-Q
-s, --silent, --quiet
--taskmastertrace=FILE
--tree=OPTIONS
.EE

.IP "" 6
Any other SCons command-line options that are specified
do not cause errors
but have no effect on the
.B build
command
(mainly because they affect how the SConscript files are read,
which only happens once at the beginning of interactive mode).

.TP 6
.BI clean "[OPTIONS] [TARGETS] ..."
Cleans the specified
.I TARGETS
(and their dependencies)
with the specified options.
.B c
is a synonym.
This command is itself a synonym for
.B "build --clean"

.TP 6
.BI exit
Exits SCons interactive mode.
You can also exit by terminating input
(CTRL+D on UNIX or Linux systems,
CTRL+Z on Windows systems).

.TP 6
.BI help "[COMMAND]"
Provides a help message about
the commands available in SCons interactive mode.
If
.I COMMAND
is specified,
.B h
and
.B ?
are synonyms.

.TP 6
.BI shell "[COMMANDLINE]"
Executes the specified
.I COMMANDLINE
in a subshell.
If no
.I COMMANDLINE
is specified,
executes the interactive command interpreter
specified in the
.B SHELL
environment variable
(on UNIX and Linux systems)
or the
.B COMSPEC
environment variable
(on Windows systems).
.B sh
and
.B !
are synonyms.

.TP 6
.B version
Prints SCons version information.
.RE

.IP
An empty line repeats the last typed command.
Command-line editing can be used if the
.B readline
module is available.

.ES
$ scons --interactive
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons>>> build -n prog
scons>>> exit
.EE

.TP
.RI -j " N" ", --jobs=" N
Specifies the number of jobs (commands) to run simultaneously.
If there is more than one
.B -j
option, the last one is effective.
.\" ??? If the
.\" .B -j
.\" option
.\" is specified without an argument,
.\" .B scons
.\" will not limit the number of
.\" simultaneous jobs.

.TP
-k, --keep-going
Continue as much as possible after an error.  The target that
failed and those that depend on it will not be remade, but other
targets specified on the command line will still be processed.

.\" .TP
.\" .RI  -l " N" ", --load-average=" N ", --max-load=" N
.\" No new jobs (commands) will be started if
.\" there are other jobs running and the system load
.\" average is at least
.\" .I N
.\" (a floating-point number).

.\"
.\" .TP
.\" --list-derived
.\" List derived files (targets, dependencies) that would be built,
.\" but do not build them.
.\" [XXX This can probably go away with the right
.\" combination of other options.  Revisit this issue.]
.\"
.\" .TP
.\" --list-actions
.\" List derived files that would be built, with the actions
.\" (commands) that build them.  Does not build the files.
.\" [XXX This can probably go away with the right
.\" combination of other options.  Revisit this issue.]
.\"
.\" .TP
.\" --list-where
.\" List derived files that would be built, plus where the file is
.\" defined (file name and line number).  Does not build the files.
.\" [XXX This can probably go away with the right
.\" combination of other options.  Revisit this issue.]

.TP
-m
Ignored for compatibility with non-GNU versions of
.BR make .

.TP
.RI --max-drift= SECONDS
Set the maximum expected drift in the modification time of files to
.IR SECONDS .
This value determines how long a file must be unmodified
before its cached content signature
will be used instead of
calculating a new content signature (MD5 checksum)
of the file's contents.
The default value is 2 days, which means a file must have a
modification time of at least two days ago in order to have its
cached content signature used.
A negative value means to never cache the content
signature and to ignore the cached value if there already is one. A value
of 0 means to always use the cached signature,
no matter how old the file is.

.TP
.RI --md5-chunksize= KILOBYTES
Set the block size used to compute MD5 signatures to
.IR KILOBYTES . 
This value determines the size of the chunks which are read in at once when
computing MD5 signatures.  Files below that size are fully stored in memory
before performing the signature computation while bigger files are read in
block-by-block. A huge block-size leads to high memory consumption while a very
small block-size slows down the build considerably.

The default value is to use a chunk size of 64 kilobytes, which should
be appropriate for most uses.

.TP
-n, --just-print, --dry-run, --recon
No execute.  Print the commands that would be executed to build
any out-of-date target files, but do not execute the commands.

.TP
.RI --no-site-dir
Prevents the automatic addition of the standard
.I site_scons
dirs to
.IR sys.path .
Also prevents loading the
.I site_scons/site_init.py
modules if they exist, and prevents adding their
.I site_scons/site_tools
dirs to the toolpath.

.\" .TP
.\" .RI -o " file" ", --old-file=" file ", --assume-old=" file
.\" Do not rebuild
.\" .IR file ,
.\" and do
.\" not rebuild anything due to changes in the contents of
.\" .IR file .
.\" .TP
.\" .RI --override " file"
.\" Read values to override specific build environment variables
.\" from the specified
.\" .IR file .
.\" .TP
.\" -p
.\" Print the data base (construction environments,
.\" Builder and Scanner objects) that are defined
.\" after reading the SConscript files.
.\" After printing, a normal build is performed
.\" as usual, as specified by other command-line options.
.\" This also prints version information
.\" printed by the
.\" .B -v
.\" option.
.\"
.\" To print the database without performing a build do:
.\"
.\" .ES
.\" scons -p -q
.\" .EE

.TP
.RI --profile= file
Run SCons under the Python profiler
and save the results in the specified
.IR file .
The results may be analyzed using the Python
pstats module.

.TP
-q, --question
Do not run any commands, or print anything.  Just return an exit
status that is zero if the specified targets are already up to
date, non-zero otherwise.
.TP
-Q
Quiets SCons status messages about
reading SConscript files,

src/script/scons.py
#! /usr/bin/env python
#
# SCons - a Software Constructor
#
# __COPYRIGHT__
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
# KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

__revision__ = "__FILE__ __REVISION__ __DATE__ __DEVELOPER__"

__version__ = "__VERSION__"

__build__ = "__BUILD__"

__buildsys__ = "__BUILDSYS__"

__date__ = "__DATE__"

__developer__ = "__DEVELOPER__"

import os
import sys

##############################################################################
# BEGIN STANDARD SCons SCRIPT HEADER
#
# This is the cut-and-paste logic so that a self-contained script can
# interoperate correctly with different SCons versions and installation
# locations for the engine.  If you modify anything in this section, you
# should also change other scripts that use this same header.
##############################################################################

# Strip the script directory from sys.path() so on case-insensitive
# (WIN32) systems Python doesn't think that the "scons" script is the
# "SCons" package.  Replace it with our own library directories
# (version-specific first, in case they installed by hand there,
# followed by generic) so we pick up the right version of the build
# engine modules if they're in either directory.


# Check to see if the python version is > 3.0 which is currently unsupported
# If so exit with error message
try:
    if  sys.version_info >= (3,0,0):
        msg = "scons: *** SCons version %s does not run under Python version %s.\n\
Python 3.0 and later are not yet supported.\n"
        sys.stderr.write(msg % (__version__, sys.version.split()[0]))
        sys.exit(1)
except AttributeError:
    # Pre-1.6 Python has no sys.version_info
    # No need to check version as we then know the version is < 3.0.0 and supported
    pass

script_dir = sys.path[0]

if script_dir in sys.path:
    sys.path.remove(script_dir)

libs = []

if "SCONS_LIB_DIR" in os.environ:
    libs.append(os.environ["SCONS_LIB_DIR"])

local_version = 'scons-local-' + __version__
local = 'scons-local'
if script_dir:
    local_version = os.path.join(script_dir, local_version)
    local = os.path.join(script_dir, local)
libs.append(os.path.abspath(local_version))
libs.append(os.path.abspath(local))

scons_version = 'scons-%s' % __version__

# preferred order of scons lookup paths
prefs = []

try:
    import pkg_resources
except ImportError:
    pass
else:
    # when running from an egg add the egg's directory 
    try:
        d = pkg_resources.get_distribution('scons')
    except pkg_resources.DistributionNotFound:
        pass
    else:
        prefs.append(d.location)

if sys.platform == 'win32':
    # sys.prefix is (likely) C:\Python*;
    # check only C:\Python*.
    prefs.append(sys.prefix)
    prefs.append(os.path.join(sys.prefix, 'Lib', 'site-packages'))
else:
    # On other (POSIX) platforms, things are more complicated due to
    # the variety of path names and library locations.  Try to be smart
    # about it.
    if script_dir == 'bin':
        # script_dir is `pwd`/bin;
        # check `pwd`/lib/scons*.
        prefs.append(os.getcwd())
    else:
        if script_dir == '.' or script_dir == '':
            script_dir = os.getcwd()
        head, tail = os.path.split(script_dir)
        if tail == "bin":
            # script_dir is /foo/bin;
            # check /foo/lib/scons*.
            prefs.append(head)

    head, tail = os.path.split(sys.prefix)
    if tail == "usr":
        # sys.prefix is /foo/usr;
        # check /foo/usr/lib/scons* first,
        # then /foo/usr/local/lib/scons*.
        prefs.append(sys.prefix)
        prefs.append(os.path.join(sys.prefix, "local"))
    elif tail == "local":
        h, t = os.path.split(head)
        if t == "usr":
            # sys.prefix is /foo/usr/local;
            # check /foo/usr/local/lib/scons* first,
            # then /foo/usr/lib/scons*.
            prefs.append(sys.prefix)
            prefs.append(head)
        else:
            # sys.prefix is /foo/local;
            # check only /foo/local/lib/scons*.
            prefs.append(sys.prefix)
    else:
        # sys.prefix is /foo (ends in neither /usr or /local);
        # check only /foo/lib/scons*.
        prefs.append(sys.prefix)

    temp = [os.path.join(x, 'lib') for x in prefs]
    temp.extend([os.path.join(x,
                                           'lib',
                                           'python' + sys.version[:3],
                                           'site-packages') for x in prefs])
    prefs = temp

    # Add the parent directory of the current python's library to the
    # preferences.  On SuSE-91/AMD64, for example, this is /usr/lib64,
    # not /usr/lib.
    try:
        libpath = os.__file__
    except AttributeError:
        pass
    else:
        # Split /usr/libfoo/python*/os.py to /usr/libfoo/python*.
        libpath, tail = os.path.split(libpath)
        # Split /usr/libfoo/python* to /usr/libfoo
        libpath, tail = os.path.split(libpath)
        # Check /usr/libfoo/scons*.
        prefs.append(libpath)

# Look first for 'scons-__version__' in all of our preference libs,
# then for 'scons'.
libs.extend([os.path.join(x, scons_version) for x in prefs])
libs.extend([os.path.join(x, 'scons') for x in prefs])

sys.path = libs + sys.path

##############################################################################
# END STANDARD SCons SCRIPT HEADER
##############################################################################

if __name__ == "__main__":
    import SCons.Script
    # this does all the work, and calls sys.exit
    # with the proper exit status when done.
    SCons.Script.main()

# Local Variables:
# tab-width:4
# indent-tabs-mode:nil
# End:
# vim: set expandtab tabstop=4 shiftwidth=4:
